### PR TITLE
Fixes #12729 Flatten Case expression

### LINKED ISSUE #12729: Flatten CASE expressions to avoid error "Case expressions may only be nested to level 10"
This fails on SQL server, but works for in-memory/SQLite.

```
System.Data.SqlClient.SqlException (0x80131904): Case expressions may only be nested to level 10.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReSystem.Data.SqlClient.SqlException: Case expressions may only be nested to level 10.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData()
   at System.Data.SqlClient.SqlDataReader.get_MetaData()
   at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString)
   at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, SqlDataReader ds)
   at System.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, TaskCompletionSource`1 completion, Int32 timeout, Task& task, Boolean asyncWrite, String method)
   at System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior)
   at System.Data.SqlClient.SqlCommand.ExecuteDbDataReader(CommandBehavior behavior)
   at System.Data.Common.DbCommand.ExecuteReader()
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, DbCommandMethod executeMethod, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteReader(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.BufferlessMoveNext(DbContext _, Boolean buffer)
   at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider._TrackEntities[TOut,TIn](IEnumerable`1 results, QueryContext queryContext, IList`1 entityTrackingInfos, IList`1 entityAccessors)+MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
   at System.Collections.Generic.List`1.AddEnumerable(IEnumerable`1 enumerable)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
```

### Steps to reproduce

```c#
using System;
using System.Linq;
using System.Threading.Tasks;

using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Console;

namespace EFCoreBugs
{
    class Program
    {
        static async Task Main(string[] args)
        {
            using (var db = new TestDb())
            {
                await db.Database.EnsureDeletedAsync();
                await db.Database.EnsureCreatedAsync();
            }
            using (var db = new TestDb())
            {
                var key = Guid.NewGuid();
                db.Todos
                    .Where(x => key ==
                        (x.Type == TodoType.foo9
                            ? key
                            : x.Type == TodoType.foo3
                                ? key
                                : x.Type == TodoType.foo1
                                    ? key
                                    : x.Type == TodoType.foo4
                                        ? key
                                        : x.Type == TodoType.foo6
                                            ? key
                                            : x.Type == TodoType.foo7
                                                ? key
                                                : x.Type == TodoType.foo8
                                                    ? key
                                                    : x.Type == TodoType.foo10
                                                        ? key
                                                        : x.Type == TodoType.foo11
                                                            ? key
                                                            : x.Type == TodoType.foo4
                                                                ? key
                                                                : x.Type == TodoType.foo2
                                                                    ? key
                                                                    : key))
                    .ToList();
            }

            Console.ReadLine();
        }
    }

    public class TestDb : DbContext
    {
        public DbSet<Todo> Todos { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            var log = new LoggerFactory(new[] { new ConsoleLoggerProvider((_, __) => __ == LogLevel.Error, true) });
            optionsBuilder.UseLoggerFactory(log);
            optionsBuilder.UseSqlServer("Data Source=(localdb)\\MSSQLLocalDB;Initial Catalog=master;Database=bugtest;Integrated Security=True;Connect Timeout=30;Encrypt=False;TrustServerCertificate=True;ApplicationIntent=ReadWrite;MultiSubnetFailover=False");
        }
    }

    public class Todo
    {
        public Guid Id { get; set; }
        public TodoType Type { get; set; }
    }

    public enum TodoType
    {
        foo0 = 0,
        foo1 = 1,
        foo2 = 2,
        foo3 = 3,
        foo4 = 4,
        foo6 = 6,
        foo7 = 7,
        foo8 = 8,
        foo9 = 9,
        foo10 = 10,
        foo11 = 11
    }
}
```

### Further technical details
EF Core version: 2.1.2
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: win10 x64
IDE:  Visual Studio 2017 15.8.0 Preview 4.0

### PR DESCRIPTION
**Pull request by ptjhuang**
Summary of the changes
- Added CaseWhenFlatteningExpressionVisitor postprocessor 
- Recursively collapse relevant CaseExpression
Fixes #12729

### CODE REVIEW

**Comment by ajcvickers** [comment]
@ptjhuang Thanks for submitting a PR. Assigning to @smitpatel for initial review.

File: src/EFCore.Relational/Query/Internal/CaseWhenFlatteningExpressionVisitor.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+        public CaseWhenFlatteningExpressionVisitor(ISqlExpressionFactory sqlExpressionFactory)
+        {
+            _sqlExpressionFactory = sqlExpressionFactory;
+        }
+
+        protected override Expression VisitExtension(Expression node)
+        {
+            // Only applies to 'CASE WHEN condition...' not 'CASE operand WHEN...'
+            if (node is CaseExpression caseExpr && caseExpr.Operand == null)
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Internal/CaseWhenFlatteningExpressionVisitor.cs]: use full variable name as `caseExpression` & `nestedCaseExpression`

File: test/EFCore.Specification.Tests/Query/SimpleQueryTestBase.Select.cs [hunk 2]
```diff
@@ -735,6 +735,28 @@ public virtual Task Select_conditional_with_null_comparison_in_test(bool async)
...
 
+        [ConditionalTheory]
+        [MemberData(nameof(IsAsyncData))]
+        public virtual Task Select_over_10_nested_ternary_condition(bool isAsync)
+        {
+            return AssertQuery(
+                isAsync,
+                os => from c in os.Set<Customer>()
+                      select
+                      c.CustomerID == "1" ? "01" :
```
Comment (smitpatel) [inline test/EFCore.Specification.Tests/Query/SimpleQueryTestBase.Select.cs]: Formatting scheme we have is generally like
```C#
c.CustomerID == "1"
    ? "01"
    : c.CustomerID == "2"
        ? "02"
        : null
```

File: src/EFCore.Relational/Query/Internal/CaseWhenFlatteningExpressionVisitor.cs (hunk 1 above)
Comment (ptjhuang) [reply src/EFCore.Relational/Query/Internal/CaseWhenFlatteningExpressionVisitor.cs]: Done

File: test/EFCore.Specification.Tests/Query/SimpleQueryTestBase.Select.cs (hunk 2 above)
Comment (ptjhuang) [reply test/EFCore.Specification.Tests/Query/SimpleQueryTestBase.Select.cs]: Are you sure? :)  Done anyway.

**Comment by smitpatel** [comment]
@ptjhuang - Thank you for contribution.

**Comment by ptjhuang** [comment]
Any chance this could make 3.1?

**Comment by ajcvickers** [comment]
@ptjhuang Unfortunately not. 3.1 was locked down some time ago.

**Pull request merged**